File: src/sqtop/views/job_detail.py
"""Job detail modal — shows scontrol show job output."""
from __future__ import annotations

from textual import work
from textual.app import ComposeResult
from textual.binding import Binding
from textual.screen import ModalScreen
from textual.containers import ScrollableContainer
from textual.widgets import Label, Static

from .detail import DetailView
from ..slurm import fetch_job_efficiency

_TERMINAL_STATES = {"COMPLETED", "FAILED", "CANCELLED", "TIMEOUT", "NODE_FAIL", "OUT_OF_MEMORY"}
_BAR_WIDTH = 10


def _eff_bar(fraction: float, bar_width: int = _BAR_WIDTH) -> str:
    """Return a Rich-markup efficiency bar string like '[green]██████░░░░[/] 62%'."""
    pct = round(min(max(fraction, 0.0), 1.0) * 100)
    filled = round(pct / 100 * bar_width)
    bar = "█" * filled + "░" * (bar_width - filled)
    color = "green" if pct >= 70 else ("yellow" if pct >= 40 else "red")
    return f"[{color}]{bar}[/]  {pct:3}%"


def _build_efficiency_text(eff: dict) -> str:
    """Build a two-line Rich markup string for CPU and memory efficiency."""
    cpu_bar = _eff_bar(eff["cpu_eff"])
    mem_bar = _eff_bar(eff["mem_eff"])
    cpu_line = (
        f"  [bold]CPU efficiency:[/]  {cpu_bar}"
        f"   [dim](used {eff['cpu_used_str']} of {eff['cpu_alloc_str']} CPU-time)[/]"
    )
    mem_line = (
        f"  [bold]Mem efficiency:[/]  {mem_bar}"
        f"   [dim](peak {eff['mem_peak_mb']} MB of {eff['mem_alloc_mb']} MB allocated)[/]"
    )
    return cpu_line + "\n" + mem_line


class JobDetailScreen(ModalScreen[None]):
    """Modal that displays pre-fetched job detail via DetailView."""

    BINDINGS = [
        Binding("escape", "dismiss(None)", show=False),
        Binding("q", "dismiss(None)", "Close", show=True),
    ]

    CSS = """
    JobDetailScreen { align: center middle; }
    #job-detail-dialog {
        width: 90%; height: 80%;
        border: double $primary;
        background: $surface;
        padding: 0;
    }
    #job-detail-title {
        text-style: bold;
        padding: 0 2;
        margin-bottom: 1;
        background: $primary;
        color: $background;
        width: 100%;
    }
    #job-detail-efficiency {
        padding: 1 2 0 2;
    }
    #job-detail-scroll {
        height: 1fr;
        padding: 0 2;
    }
    """

    def __init__(self, job_id: str, data: dict[str, str]) -> None:
        super().__init__()
        self._job_id = job_id
        self._data = data

    def compose(self) -> ComposeResult:
        name = self._data.get("JobName", "")
        header = f"Job {self._job_id}" + (f" — {name}" if name else "")
        with Static(id="job-detail-dialog"):
            yield Label(header, id="job-detail-title")
            yield Static("", id="job-detail-efficiency")
            with ScrollableContainer(id="job-detail-scroll"):
                yield DetailView(id="job-detail-view")

    def on_mount(self) -> None:
        self.call_after_refresh(self._show_content)

    def _show_content(self) -> None:
        self.query_one("#job-detail-view", DetailView).show_job(self._data)
        state = self._data.get("JobState", "").upper()
        if state in _TERMINAL_STATES:
            self._load_efficiency()
        else:
            self.query_one("#job-detail-efficiency", Static).display = False

    @work(thread=True)
    def _load_efficiency(self) -> None:
        """Fetch and display efficiency bars for terminal-state jobs (background thread)."""
        eff = fetch_job_efficiency(self._job_id)
        if not eff.get("available") or (eff["mem_peak_mb"] == 0 and eff["cpu_eff"] == 0.0):
            self.app.call_from_thread(self._hide_efficiency)
            return
        text = _build_efficiency_text(eff)
        self.app.call_from_thread(self.query_one("#job-detail-efficiency", Static).update, text)

    def _hide_efficiency(self) -> None:
        self.query_one("#job-detail-efficiency", Static).display = False

File: src/views/detail/job_detail.rs
//! Job detail modal with efficiency bars.

use crate::slurm::fetch::JobEfficiency;
use crate::views::detail::Outcome;
use crossterm::event::{KeyCode, KeyEvent};
use ratatui::{
    layout::{Alignment, Constraint, Direction, Layout, Rect},
    style::{Color, Modifier, Style},
    text::{Line, Span},
    widgets::{Block, Borders, Paragraph, Wrap},
    Frame,
};
use std::collections::HashMap;

/// Job detail viewer with efficiency bars for terminal-state jobs.
pub struct JobDetailScreen {
    job_id: String,
    data: HashMap<String, String>,
    efficiency: Option<JobEfficiency>,
    scroll_offset: usize,
    lines: Vec<String>,
}

impl JobDetailScreen {
    /// Create a new job detail screen.
    pub fn new(job_id: String, data: HashMap<String, String>) -> Self {
        let lines = build_detail_lines(&job_id, &data);
        Self {
            job_id,
            data,
            efficiency: None,
            scroll_offset: 0,
            lines,
        }
    }

    /// Set the efficiency data.
    pub fn set_efficiency(&mut self, eff: JobEfficiency) {
        self.efficiency = Some(eff);
    }

    /// Handle key input.
    pub fn handle_key(&mut self, key: KeyEvent) -> Outcome {
        match key.code {
            KeyCode::Esc | KeyCode::Char('q') => Outcome::Close,
            KeyCode::Up | KeyCode::Char('k') => {
                self.scroll_offset = self.scroll_offset.saturating_sub(1);
                Outcome::None
            }
            KeyCode::Down | KeyCode::Char('j') => {
                if self.scroll_offset < self.lines.len().saturating_sub(1) {
                    self.scroll_offset += 1;
                }
                Outcome::None
            }
            KeyCode::PageUp => {
                self.scroll_offset = self.scroll_offset.saturating_sub(10);
                Outcome::None
            }
            KeyCode::PageDown => {
                let max_scroll = self.lines.len().saturating_sub(1);
                self.scroll_offset = (self.scroll_offset + 10).min(max_scroll);
                Outcome::None
            }
            _ => Outcome::None,
        }
    }

    /// Render the job detail screen.
    pub fn render(&self, f: &mut Frame, area: Rect) {
        // Center the dialog (90% width, 85% height)
        let width = (area.width * 9 / 10).clamp(60, 140);
        let height = (area.height * 85 / 100).clamp(20, 50);
        let dialog_area = centered_rect(width, height, area);

        // Render border
        let block = Block::default()
            .borders(Borders::ALL)
            .border_style(
                Style::default()
                    .fg(Color::Cyan)
                    .add_modifier(Modifier::BOLD),
            )
            .style(Style::default().bg(Color::Black));
        f.render_widget(block.clone(), dialog_area);

        // Inner layout
        let inner = block.inner(dialog_area);

        // Calculate constraints based on efficiency availability
        let mut constraints = vec![Constraint::Length(1)]; // title
        if self.efficiency.is_some() {
            constraints.push(Constraint::Length(3)); // efficiency bars
        }
        constraints.push(Constraint::Min(0)); // detail content

        let chunks = Layout::default()
            .direction(Direction::Vertical)
            .constraints(constraints)
            .split(inner);

        // Title
        let name = self.data.get("JobName").map(|s| s.as_str()).unwrap_or("");
        let header_text = if name.is_empty() {
            format!("Job {}", self.job_id)
        } else {
            format!("Job {} — {}", self.job_id, name)
        };
        let title = Paragraph::new(Span::styled(
            header_text,
            Style::default()
                .fg(Color::Black)
                .bg(Color::Cyan)
                .add_modifier(Modifier::BOLD),
        ))
        .alignment(Alignment::Center);
        f.render_widget(title, chunks[0]);

        // Efficiency bars (if available)
        let content_chunk_idx = if let Some(eff) = &self.efficiency {
            if eff.available && (eff.mem_peak_mb > 0 || eff.cpu_eff > 0.0) {
                let eff_text = build_efficiency_text(eff);
                let eff_para = Paragraph::new(eff_text);
                f.render_widget(eff_para, chunks[1]);
                2
            } else {
                1
            }
        } else {
            1
        };

        // Detail content
        let content_height = chunks[content_chunk_idx].height as usize;
        let visible_lines: Vec<Line> = self
            .lines
            .iter()
            .skip(self.scroll_offset)
            .take(content_height)
            .map(|s| Line::from(Span::raw(s.as_str())))
            .collect();

        let content = Paragraph::new(visible_lines).wrap(Wrap { trim: false });
        f.render_widget(content, chunks[content_chunk_idx]);
    }

    /// Get the plain text content for clipboard copy.
    pub fn plain_text(&self) -> String {
        self.lines.join("\n")
    }

    /// Get the label for clipboard copy.
    pub fn label(&self) -> String {
        format!("Job {} Detail", self.job_id)
    }
}

fn build_detail_lines(_job_id: &str, data: &HashMap<String, String>) -> Vec<String> {
    let mut lines = vec!["Job Detail\n".to_string()];

    // Highlight keys (shown first)
    let highlight_keys = [
        "JobId",
        "JobName",
        "UserId",
        "JobState",
        "NumNodes",
        "NumCPUs",
        "TimeLimit",
        "SubmitTime",
        "StartTime",
        "EndTime",
        "Partition",
        "NodeList",
        "Reason",
        "Priority",
    ];

    for key in &highlight_keys {
        if let Some(value) = data.get(*key) {
            lines.push(format!("  {}: {}", key, value));
        }
    }

    // Other keys
    for (k, v) in data {
        if !highlight_keys.contains(&k.as_str()) {
            lines.push(format!("  {}: {}", k, v));
        }
    }

    lines
}

fn build_efficiency_text(eff: &JobEfficiency) -> Vec<Line<'_>> {
    let cpu_bar = eff_bar(eff.cpu_eff);
    let mem_bar = eff_bar(eff.mem_eff);

    vec![
        Line::from(vec![
            Span::styled(
                "  CPU efficiency:  ",
                Style::default().add_modifier(Modifier::BOLD),
            ),
            cpu_bar,
            Span::styled(
                format!(
                    "   (used {} of {} CPU-time)",
                    eff.cpu_used_str, eff.cpu_alloc_str
                ),
                Style::default().fg(Color::DarkGray),
            ),
        ]),
        Line::from(vec![
            Span::styled(
                "  Mem efficiency:  ",
                Style::default().add_modifier(Modifier::BOLD),
            ),
            mem_bar,
            Span::styled(
                format!(
                    "   (peak {} MB of {} MB allocated)",
                    eff.mem_peak_mb, eff.mem_alloc_mb
                ),
                Style::default().fg(Color::DarkGray),
            ),
        ]),
    ]
}

fn eff_bar(fraction: f64) -> Span<'static> {
    const BAR_WIDTH: usize = 10;
    let pct = (fraction.clamp(0.0, 1.0) * 100.0).round() as u32;
    let filled = ((pct as f64 / 100.0) * BAR_WIDTH as f64).round() as usize;
    let bar = "█".repeat(filled) + &"░".repeat(BAR_WIDTH - filled);

    let color = if pct >= 70 {
        Color::Green
    } else if pct >= 40 {
        Color::Yellow
    } else {
        Color::Red
    };

    Span::styled(format!("{}  {:3}%", bar, pct), Style::default().fg(color))
}

fn centered_rect(width: u16, height: u16, area: Rect) -> Rect {
    let x = area.x + (area.width.saturating_sub(width)) / 2;
    let y = area.y + (area.height.saturating_sub(height)) / 2;
    Rect {
        x,
        y,
        width: width.min(area.width),
        height: height.min(area.height),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_detail_lines_includes_highlight_keys() {
        let mut data = HashMap::new();
        data.insert("JobId".to_string(), "12345".to_string());
        data.insert("JobName".to_string(), "test_job".to_string());
        data.insert("JobState".to_string(), "RUNNING".to_string());

        let lines = build_detail_lines("12345", &data);
        assert!(lines.iter().any(|l| l.contains("JobId: 12345")));
        assert!(lines.iter().any(|l| l.contains("JobName: test_job")));
        assert!(lines.iter().any(|l| l.contains("JobState: RUNNING")));
    }

    #[test]
    fn test_eff_bar_green_above_70() {
        let span = eff_bar(0.75);
        // The bar should contain 8 filled chars and 2 empty for 75%
        assert!(span.content.contains("██████"));
    }

    #[test]
    fn test_eff_bar_red_below_40() {
        let span = eff_bar(0.25);
        // The bar should contain 3 filled chars and 7 empty for 25%
        assert!(span.content.contains("███░"));
    }

    #[test]
    fn test_set_efficiency_updates_screen_state() {
        use crate::slurm::fetch::JobEfficiency;
        let data = HashMap::new();
        let mut screen = JobDetailScreen::new("12345".to_string(), data);

        let eff = JobEfficiency {
            available: true,
            cpu_eff: 0.85,
            mem_eff: 0.65,
            cpu_used_str: "1:23:45".to_string(),
            cpu_alloc_str: "1:38:40".to_string(),
            mem_peak_mb: 650,
            mem_alloc_mb: 1000,
        };

        screen.set_efficiency(eff.clone());
        assert!(screen.efficiency.is_some());
        assert_eq!(screen.efficiency.as_ref().unwrap().cpu_eff, 0.85);
        assert_eq!(screen.efficiency.as_ref().unwrap().mem_eff, 0.65);
    }
}
